### PR TITLE
Show supported languages for voice search

### PR DESCRIPTION
**Pull request by felipeerias**
This PR moves the responsibility of returning the list of supported languages to the adapter class for each voice recognition service (`MKSpeechRecognizer` and `HVRSpeechRecognizer`).

The Voice Search setting used to display the whole list of languages supported by Wolvic. However, many of those languages were not actually supported by the services that we use on each platform.

Additional changes to voice recognition:

- Add a callback method `onCanceled()` for when voice recognition is canceled. This corresponds to MeetKai's `CANCELED` state.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -576,12 +576,12 @@ dependencies {
...
-    hvrImplementation 'com.huawei.hms:ml-computer-voice-asr:3.1.0.300'
-    hvr3dofImplementation 'com.huawei.hms:ml-computer-voice-asr:3.1.0.300'
-    hvrImplementation 'com.huawei.hms:location:6.2.0.300'
-    hvr3dofImplementation 'com.huawei.hms:location:6.2.0.300'
+    hvrImplementation 'com.huawei.hms:hianalytics:6.5.0.300'
+    hvr3dofImplementation 'com.huawei.hms:hianalytics:6.5.0.300'
+    hvrImplementation 'com.huawei.hms:ml-computer-voice-asr:3.5.0.301'
+    hvr3dofImplementation 'com.huawei.hms:ml-computer-voice-asr:3.5.0.301'
+    hvrImplementation 'com.huawei.hms:location:6.4.0.300'
+    hvr3dofImplementation 'com.huawei.hms:location:6.4.0.300'
```
Comment (svillar) [inline app/build.gradle:584]: Are these changes strictly required?

You're using 6.4 for location services. Is that correct? I just ask because you're using 6.5 for the others.

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/VoiceSearchLanguageOptionsView.java [hunk 2]
```diff
@@ -32,11 +34,17 @@ private void initialize(Context aContext) {
         updateUI();
     }
 
+    private SpeechRecognizer getSpeechRecognizer() {
+        VRBrowserApplication application = (VRBrowserApplication) getContext().getApplicationContext();
+        return application.getSpeechRecognizer();
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/VoiceSearchLanguageOptionsView.java:39]: You can use a single line for this, but not a strong opinion.

File: app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java [hunk 3]
```diff
@@ -6,21 +6,58 @@
...
+        mRecognizer = MLAsrRecognizer.createAsrRecognizer(mContext);
+        mRecognizer.getLanguages(new MLAsrRecognizer.LanguageCallback() {
+            @Override
+            public void onResult(List<String> list) {
+                setSupportedLanguages(list);
+            }
+
+            @Override
+            public void onError(int i, String s) {
+            }
```
Comment (svillar) [inline app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java:49]: We should perhaps add some debugging statement here.

File: app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java [hunk 4]
```diff
@@ -6,21 +6,58 @@
...
+        });
+    }
+
+    private synchronized void setSupportedLanguages(List<String> supportedLanguages) {
+        mSupportedLanguages.clear();
+        mSupportedLanguages.add(LocaleUtils.DEFAULT_LANGUAGE_ID);
+        mSupportedLanguages.addAll(supportedLanguages);
```
Comment (svillar) [inline app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java:56]: Is it guaranteed that DEFAULT_LANGUAGE_ID is not included in supportedLanguages? Otherwise we would end up with a duplicated id.

File: app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java [hunk 5]
```diff
@@ -6,21 +6,58 @@
...
+                setSupportedLanguages(list);
+            }
+
+            @Override
+            public void onError(int i, String s) {
+            }
+        });
+    }
+
+    private synchronized void setSupportedLanguages(List<String> supportedLanguages) {
```
Comment (svillar) [inline app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java:53]: This is only called from the constructor, why do we need to make it synchronized?

File: app/src/common/shared/com/igalia/wolvic/speech/SpeechRecognizer.java [hunk 6]
```diff
@@ -40,4 +40,7 @@ interface Callback {
     boolean shouldDisplayStoreDataPrompt();
     default boolean supportsASR(@NonNull Settings settings) {return true;}
     boolean isSpeechError(int code);
+    String[] getSupportedLanguagesNames();
+    int getIndexForLanguageId(String languageId);
+    String getLanguageIdForIndex(int checkedRadioButtonId);
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/speech/SpeechRecognizer.java]: I think we could live with it temporarily because it follows the same structure as the current code but I strongly disagree with having these 2 methods in the speech recognizers. Those two methods are used to associate languages with visual elements (radio buttons in this case). This should be something that  must be done at a very different (higher) level. The settings widget should be the one getting the list of supported languages and building and controlling the UI to select the language instead of relying on some indexes from some random array in the speech recognizers. 

Please file a bug report and do add a FIXME linking to that issue here so that we don't forget,

**Review by svillar — CHANGES_REQUESTED** [review]
This is great stuff. I have some comments though.

BTW could you please unify the first and third commits, otherwise it's a bit confusing because you make some changes and then you modify those changes so it's difficult to compare with the current code because there are 2 iterations of changes on top.

File: app/build.gradle (hunk 1 above)
Comment (felipeerias) [reply app/build.gradle:584]: Moved to a separate PR.

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/VoiceSearchLanguageOptionsView.java (hunk 2 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/settings/VoiceSearchLanguageOptionsView.java:39]: Done.

File: app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java (hunk 4 above)
Comment (felipeerias) [reply app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java:58]: Yes, `DEFAULT_LANGUAGE_ID` is just the string `"default"` which is not a valid language/locale.

File: app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java (hunk 3 above)
Comment (felipeerias) [reply app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java:51]: Done.

File: app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java (hunk 5 above)
Comment (felipeerias) [reply app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java:55]: I was trying to avoid a possible error because fetching the list of languages happens asynchronously. However, it doesn't look possible to have a concurrency problem because the call to fetch the languages is only invoked once, in the constructor.

I will remove this.

File: app/src/common/shared/com/igalia/wolvic/speech/SpeechRecognizer.java (hunk 6 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/speech/SpeechRecognizer.java]: Good point. I opened https://github.com/Igalia/wolvic/issues/151

File: app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java (hunk 4 above)
Comment (svillar) [reply app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java:59]: My question was more focused on whether the code prevents that from happening, not only for default but for any other language. I think what I really meant is that mSupportedLanguages should be a Set instead of a List

File: app/src/common/shared/com/igalia/wolvic/speech/MKSpeechRecognizer.java [hunk 7]
```diff
@@ -30,11 +36,26 @@ public class MKSpeechRecognizer implements SpeechRecognizer, ISpeechRecognitionL
     // temporary
     private static float FINAL_CONF = 1.0f;
 
+    private static final List<String> mSupportedLanguages = Arrays.asList(
+            LocaleUtils.DEFAULT_LANGUAGE_ID,
+            "en-US", "zh-CN", "ja-JP", "fr-FR", "de-DE", "es-ES", "ru-RU", "ko-KR",
+            "it-IT", "pl-PL", "sv-SE", "fi-FI", "ar-SA", "id-ID", "th-TH", "he-IL");
+
+    private static final List<String> mSupportedLanguagesNames = new ArrayList<>();
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/speech/MKSpeechRecognizer.java:44]: Sorry for not spotting this before but as we don't allow duplicates we should use a Set instead of a List. That would save us some headaches just in case we eventually fetch those from some service and end up with duplicates somehow.

After a second thought I realized that you have to do operations with the indexes of the Array so I guess using sets is not an option. Anyway food for thought for eventual followups.

File: app/build.gradle [hunk 8]
```diff
@@ -576,12 +576,12 @@ dependencies {
...
-    hvrImplementation 'com.huawei.hms:ml-computer-voice-asr:3.5.0.301'
-    hvr3dofImplementation 'com.huawei.hms:ml-computer-voice-asr:3.5.0.301'
-    hvrImplementation 'com.huawei.hms:location:6.4.0.300'
-    hvr3dofImplementation 'com.huawei.hms:location:6.4.0.300'
+    hvrImplementation 'com.huawei.hms:hianalytics:6.3.2.300'
+    hvr3dofImplementation 'com.huawei.hms:hianalytics:6.3.2.300'
+    hvrImplementation 'com.huawei.hms:ml-computer-voice-asr:3.1.0.300'
+    hvr3dofImplementation 'com.huawei.hms:ml-computer-voice-asr:3.1.0.300'
+    hvrImplementation 'com.huawei.hms:location:6.2.0.300'
+    hvr3dofImplementation 'com.huawei.hms:location:6.2.0.300'
```
Comment (svillar) [inline app/build.gradle:584]: Why this? And why mixing 6.2 with 6.3.2 versions?

**Review by svillar — CHANGES_REQUESTED** [review]
I'm fine with the patch. I'm just marking it for "request changes" because I'd like to ask for the following. Could we have two different commits for the whole thing? The first one would be a small fix for the cancelled case, and the second one would be the one adding supported languages. Let me know if this is too complex, so that we would just simply squash and merge.

File: app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java (hunk 4 above)
Comment (felipeerias) [reply app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java:59]: The problem is that at the moment we need a convenient way to access items at random positions in `mSupportedLanguages`. A class like `LinkedHashSet` provides predictable iteration but not random access.

In any case, I don't think that there is a big risk of getting duplicated values. When this class is created, the content of `mSupportedLanguages` is just `{ "default", "en-US" }` while we ask the voice recognition library for its supported locales.

When the voice recognition library returns its list of supported locales, the content of `mSupportedLanguages` becomes `"default"` plus those locales.

File: app/src/common/shared/com/igalia/wolvic/speech/MKSpeechRecognizer.java (hunk 7 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/speech/MKSpeechRecognizer.java:44]: This is related to the problem that the speech recognizer code is too tied to the UI. We can fix it as part of https://github.com/Igalia/wolvic/issues/151
